active/0000-engines.md

```diff
@@ -0,0 +1,411 @@
+  like resources in other addons, but without the possibility of namespace
+  collisions.
+
+## Detailed design
+
+Engines are very similar to regular applications: they can be developed in
+isolation in Ember CLI, include addons, and contain all the same elements,
+including routes, components, initializers, etc. The primary differences are
+that an engine does not boot itself and an engine does not control the router.
+
+### Engine internals
+
+New `Engine` and `EngineInstance` classes will be introduced.
+
+Applications and engines will share ancestry. It remains TBD whether
+applications will subclass engines, or whether a common ancestor will be
+introduced.
+
+Engines and applications will share the same pattern for registry / container
+ownership and encapsulation. Both will also have initializers and instance
+initializers.
+
+Engine instances will have access to their parent instances. An engine's parent
+could be either an application or engine.
+
+#### Routable vs. route-less engines
+
+Routable engines will define their routes in a new `Ember.Routes` class. This
+class will encapsulate the functionality provided by `Router#map`, and will be
+used internally by `Ember.Router` as well (with no public interface changes of
+course).
+
+Route-less engines do not define routing maps nor can they contain routes.
+
+### Developing engines
+
+Engines can be developed in isolation as Ember CLI addon projects or as part of
+a parent application.
+
+#### Engines as addons
+
+Engines can be created as separate addon projects with:
+
+```
+ember engine <engine-name>
+```
+
+This will create a special form of an ember addon. The file structure will match
+that of a standard addon, but will have an `engine` directory instead of an
+`addon` directory.
+
+Engines can be unit tested and can also be integration tested within a dummy
+app, just like standard addons.
+
+#### In-repo engines
+
+An engine can be created within an existing application's project using a
+special `in-repo-engine` generator (similar to the `in-repo-addon` generator):
+
+```
+ember g in-repo-engine <engine-name>
+```
+
+In-repo engines can be unit tested in isolation or integration testing with the
+main application (instead of a dummy application).
+
+> Note: In-repo addons currently are created in the `/lib` directory (e.g.
+`/lib/my-addon`). Unit tests and integration tests are currently co-mingled with
+tests for the main application. It's recommended that in-repo engines provide
+better test separation than is provided for regular addons, and perhaps the
+whole in-repo addon directory structure should be re-examined at the same time
+in-repo engines are introduced.
+
+#### Engine directory structure
+
+An engine's directory will contain a file structure identical to the `app`
+directory in a standard ember-cli application, with the following exceptions:
+
+* `engine.js` instead of `app.js` - defines the `Engine` class and
```

Comment:
I'm of the personal opinion that neither in-repo addons nor in-repo engines
are really all that worth it - but that's just me.  I'm rather opinionated
on the idea of breaking up applications into things that can be worked on
by totally independent teams (That's not to suggest that the teams
themselves should be completely isolated from each-other!)

But seriously, Dan, well done on this RFC update.  I'm really on board with
this approach.  It's very similar to what I imagined in my head.  Thanks so
much for your hard work on this.

On Fri, Sep 4, 2015 at 12:39 PM, Dan Gebhardt <notifications@github.com>
wrote:

> In active/0000-engines.md
> <https://github.com/emberjs/rfcs/pull/10#discussion_r38717473>:
>
> > +In-repo engines can be unit tested in isolation or integration testing with the
> > +main application (instead of a dummy application).
> > +
> > +> Note: In-repo addons currently are created in the `/lib` directory (e.g.
> > +`/lib/my-addon`). Unit tests and integration tests are currently co-mingled with
> > +tests for the main application. It's recommended that in-repo engines provide
> > +better test separation than is provided for regular addons, and perhaps the
> > +whole in-repo addon directory structure should be re-examined at the same time
> > +in-repo engines are introduced.
> > +
> > +#### Engine directory structure
> > +
> > +An engine's directory will contain a file structure identical to the `app`
> > +directory in a standard ember-cli application, with the following exceptions:
> > +
> > +* `engine.js` instead of `app.js` - defines the `Engine` class and
>
> There's an example engine.js in the engine / parent dependencies
> <https://github.com/tomdale/rfcs/blob/master/active/0000-engines.md#engine--parent-dependencies>
> section. Besides declaring dependencies, I don't envision much difference
> from app.js.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/emberjs/rfcs/pull/10/files#r38717473>.
>